### PR TITLE
Adds custom age limits to each species

### PR DESCRIPTION
**Pull request by Henri215**
## What Does This PR Do

There are now different age limits depending on your species.
## Why It's Good For The Game

the following situations are odd to see:
* 85 years old Greys walking around when their lore talks of a 50 years life span.
Hopefully this will give the impression of more unique species.
## Testing

- Spawned lots of species and did test their pitch by trying various ages
- Tried to make a one year human on character creation
- Used the medical/security records to change the age
## Changelog
:cl:
tweak: There are now different age limits depending on your species
/:cl:

### CODE REVIEW

**Comment by Gatchapod** [comment]
Just something I'd like to point out. Average lifespan is just that. An _average_.
I believe going 5-10 years above that number would be pretty reasonable, unless we think of average age as a retirement point.

**Comment by Henri215** [comment]
> Just something I'd like to point out. Average lifespan is just that. An _average_. I believe going 5-10 years above that number would be pretty reasonable, unless we think of average age as a retirement point.

Think i should decrease it by like..10%? The previous limit based on humans was 85, their life span is 95 years.

File: code/__DEFINES/mob_defines.dm [hunk 1]
```diff
@@ -35,8 +35,8 @@
 #define MOB_SPIRIT		(1 << 9)
 #define MOB_PLANT		(1 << 10)
 
-#define AGE_MIN 17			//youngest a character can be
```
Comment (AffectedArc07) [inline code/__DEFINES/mob_defines.dm:38]: Tweaking the max is fine, but for the love of god please keep this as 17. 

I know IPCs aren't biological and chassis age can be a different thing, but please, keep this bar where it is. Given the kind of art *some* players get I don't want **any** leeway in how people portray characters. 

TLDR - This opens a can of worms I ain't prepared to deal with.

**Comment by Gatchapod** [comment]
> > Just something I'd like to point out. Average lifespan is just that. An _average_. I believe going 5-10 years above that number would be pretty reasonable, unless we think of average age as a retirement point.
> 
> Think i should decrease it by like..10%? The previous limit based on humans was 85, their life span is 95 years.

Not sure how you exactly want to handle that. I just think letting people play elderly characters is good.

Also, AA kind of makes a valid point. While I'd love to see cutting-edge-straight-off-the-assembly-line IPCs, this risks giving people an excuse to play straight-up infantile characters. We already have to deal with vox babyspeak, let's not give players opportunity to play literal babies.

File: code/__DEFINES/mob_defines.dm (hunk 1 above)
Comment (Sirryan2002) [reply code/__DEFINES/mob_defines.dm:38]: 100% Agree, the AGE_MIN needs to be maintained at 17

File: code/__DEFINES/mob_defines.dm [hunk 2]
```diff
@@ -35,8 +35,8 @@
...
-#define AGE_MAX 85			//oldest a character can be
+#define AGE_MIN 1			//youngest a character can be
```
Comment (Sirryan2002) [inline code/__DEFINES/mob_defines.dm:38]:
```suggestion
#define AGE_MIN 17			//youngest a character can be
```

**Review by Sirryan2002 — CHANGES_REQUESTED** [review]

**Comment by Ordosian** [comment]
I think you forgot to edit plasmamen age max.. 

Also.. You set drask lifespan to 150, however they live way beyond that. I understand IC reasons for it but we do have some drask players that icly are older than 150.

**Comment by Henri215** [comment]
> I think you forgot to edit plasmamen age max..
> 
> Also.. You set drask lifespan to 150, however they live way beyond that. I understand ingame reasons for it but we do have some drask players that icly are older than 150.

Its a DRAFT because of that, i have yet to figure out ages for plasmamen, drask and diona.

File: code/modules/mob/living/carbon/human/species/machine.dm [hunk 3]
```diff
@@ -1,6 +1,7 @@
 /datum/species/machine
 	name = "Machine"
 	name_plural = "Machines"
+	max_age = 150
```
Comment (warriorstar-orion) [inline code/modules/mob/living/carbon/human/species/machine.dm:4]: The current in-game year is always 2567, and IPCs were first built in 2510. While IPCs may be able to live to 150, in-universe no extant IPC can be older than ~50. This is the only species young enough that there's actually a discrepancy between the oldest possible member and the maximum age of the species.

**Comment by OctusGit** [comment]
Vox usually can live for really long times as well. Their biologically immortal so something big would prob need to be configured for them.

**Comment by Contrabang** [comment]
I think some of these races having low max ages isnt good, like unathi, vox, nian isnt good, if players want to play old characters they should be able to, and not limited by a max age limit. (Yes I am complaining because I have a 47 year old unathi and it would completely fuck their lore if they were age limited to 40).

 I get that this can be cool for other races, by having more options, but why should we lower the max age for these characters because they were given a random age limit on the wiki? If people want their lore to be old, they should be able to, the lifespan stuff is averages and not a strict rule

File: code/modules/client/preference/preferences.dm [hunk 4]
```diff
@@ -198,6 +198,8 @@ GLOBAL_LIST_INIT(special_role_times, list( //minimum age (in days) for accounts
 			dat += "<b>Gender:</b> <a href='?_src_=prefs;preference=gender'>[active_character.gender == MALE ? "Male" : (active_character.gender == FEMALE ? "Female" : "Genderless")]</a>"
 			dat += "<br>"
 			dat += "<b>Age:</b> <a href='?_src_=prefs;preference=age;task=input'>[active_character.age]</a><br>"
+			if(active_character.age < S.min_age || active_character.age > S.max_age)
+				active_character.age = rand(S.min_age, S.max_age)
```
Comment (Contrabang) [inline code/modules/client/preference/preferences.dm:202]: Could you please use a clamp here so that a 85 year old vulp isnt set to 17, but rounded down to their max age instead?

Comment (Henri215) [reply code/modules/client/preference/preferences.dm:202]: Like this?

Comment (Contrabang) [reply code/modules/client/preference/preferences.dm]: Close I think, it should probably be changed in the back end (like /Topic() and href stuff) and not in the front end like it is currently (dat += ...)

Comment (Henri215) [reply code/modules/client/preference/preferences.dm]: Any idea where? I did that here originally to prevent abuses like selecting 300 years on drasks and swapping to human so they can have it without clicking on "age".

**Comment by Ordosian** [comment]
> Vox usually can live for really long times as well. Their biologically immortal so something big would prob need to be configured for them.

Their bodies are not immortal. Their cortical stack is what makes them practically immortal. The age limit is probably for their bodies.

**Comment by OctusGit** [comment]
> > Vox usually can live for really long times as well. Their biologically immortal so something big would prob need to be configured for them.
> 
> Their bodies are not immortal. Their cortical stack is what makes them practically immortal. The age limit is probably for their bodies.

Yeah pretty much, bodies decay but stacks are immortal. Still Vox are not "reborn" when their put into a new body, the age is always the ego of the soul which to the core is just the immortal stack 🤷

File: code/__DEFINES/mob_defines.dm (hunk 1 above)
Comment (Coolrune206) [reply code/__DEFINES/mob_defines.dm:38]: Genuine question, why is this 17 instead of 18?

File: code/modules/mob/living/carbon/human/species/_species.dm [hunk 5]
```diff
@@ -3,6 +3,11 @@
 	var/name_plural 			// Pluralized name (since "[name]s" is not always valid)
 	var/icobase = 'icons/mob/human_races/r_human.dmi'    // Normal icon set.
 
+	/// Minimum age this species can have
+	var/min_age = 17
```
Comment (Sirryan2002) [inline code/modules/mob/living/carbon/human/species/_species.dm:7]:
```suggestion
	var/min_age = AGE_MIN
```

File: code/modules/client/preference/preferences.dm [hunk 6]
```diff
@@ -197,7 +197,7 @@ GLOBAL_LIST_INIT(special_role_times, list( //minimum age (in days) for accounts
 			dat += "<h2>Identity</h2>"
 			dat += "<b>Gender:</b> <a href='?_src_=prefs;preference=gender'>[active_character.gender == MALE ? "Male" : (active_character.gender == FEMALE ? "Female" : "Genderless")]</a>"
 			dat += "<br>"
-			dat += "<b>Age:</b> <a href='?_src_=prefs;preference=age;task=input'>[active_character.age]</a><br>"
+			dat += "<b>Age:</b> <a href='?_src_=prefs;preference=age;task=input'>[clamp(active_character.age, S.min_age, S.max_age)]</a><br>"
```
Comment (Sirryan2002) [inline code/modules/client/preference/preferences.dm:200]: this is the incorrect place to clamp this value, should instead be clamping the input when it's sent to `topic()`, that way we make sure we're ensuring the displayed value is lining up with the actual value of the variable.

File: code/modules/mob/living/carbon/human/human_mob.dm [hunk 7]
```diff
@@ -1817,8 +1817,8 @@ Eyes need to have significantly high darksight to shine unless the mob has the X
 
 	return ..()
 
-/mob/living/carbon/human/proc/get_age_pitch()
-	return 1.0 + 0.5*(30 - age)/80
+/mob/living/carbon/human/proc/get_age_pitch(species_pitch = 85)
+	return 1.0 + 0.5 * (30 - age) / (0.94 * species_pitch)
```
Comment (Sirryan2002) [inline code/modules/mob/living/carbon/human/human_mob.dm:1821]: were these values just what worked? Or is there a reason for these numbers specifically besides that

Comment (Henri215) [reply code/modules/mob/living/carbon/human/human_mob.dm]: OLD PITCH

https://user-images.githubusercontent.com/77684085/235300475-499b2e5c-d03a-4d6f-b175-3d182f3dbd16.mp4

NEW PITCH

https://user-images.githubusercontent.com/77684085/235300479-88714ae2-84f1-4d3d-9014-9457ee9b3429.mp4

I changed those values assuming that they are percentages based on the maximum age (which originally was 85 years), with 80 being 94% of it and 30 being 35%.

File: code/modules/client/preference/preferences.dm (hunk 6 above)
Comment (Henri215) [reply code/modules/client/preference/preferences.dm]: This place is better?

**Review by Sirryan2002 — APPROVED** [review]
can't think of anything this will break right now, code looks good.

**Review by Sirryan2002 — APPROVED** [review]
can't think of anything this will break right now, code looks good.

**Review by farie82 — APPROVED** [review]
